crates/walrus-service/src/client/daemon/routes.rs

```diff
@@ -351,6 +425,225 @@ impl From<ClientError> for GetBlobError {
+    #[error("failed to read blob {blob_id}: {message}")]
+    #[rest_api_error(reason = "BLOB_READ_FAILED", status = ApiStatusCode::Internal)]
+    BlobReadFailed { blob_id: String, message: String },
+}
+
+/// Shared implementation for concatenating and streaming multiple blobs.
+///
+/// Takes a list of ID strings (blob IDs or object IDs), resolves them to blob IDs,
+/// and returns a streaming response with the concatenated blob data.
+async fn concat_blobs_impl<T: WalrusReadClient + Send + Sync + 'static>(
+    client: Arc<T>,
+    id_strings: Vec<String>,
+    consistency_check: ConsistencyCheckType,
+    request_headers: HeaderMap,
+    response_header_config: Arc<AggregatorResponseHeaderConfig>,
+) -> Response {
+    // Validate input
+    if id_strings.is_empty() {
+        return (
+            StatusCode::BAD_REQUEST,
+            "at least one blob ID or object ID must be provided",
+        )
+            .into_response();
+    }
+
+    // Resolve all IDs to blob IDs upfront (fail fast)
+    // Track the first blob's attribute for header population
+    let mut blob_ids = Vec::new();
+    let mut first_attribute: Option<BlobAttribute> = None;
+    for (index, id_str) in id_strings.iter().enumerate() {
+        match parse_and_resolve_id(id_str.as_str(), client.as_ref()).await {
+            Ok((blob_id, attribute)) => {
+                blob_ids.push(blob_id);
+                // Only capture the first blob's attribute
+                if index == 0 {
+                    first_attribute = attribute;
+                }
+            }
+            Err(error) => return error.into_response(),
+        }
+    }
+
+    // Create streaming response
+    let client_clone = client.clone();
+    let stream = stream::iter(blob_ids).then(move |blob_id| {
+        let client = client_clone.clone();
+        async move {
+            match client.read_blob(&blob_id, consistency_check).await {
+                Ok(blob_data) => Ok(Bytes::from(blob_data)),
+                Err(error) => {
+                    // Map ClientError to appropriate ConcatBlobError
+                    let concat_error = match error.kind() {
+                        ClientErrorKind::BlobIdDoesNotExist => ConcatBlobError::BlobNotFound {
+                            blob_id: blob_id.to_string(),
+                        },
+                        ClientErrorKind::BlobIdBlocked(_) => ConcatBlobError::BlobBlocked {
+                            blob_id: blob_id.to_string(),
+                        },
+                        ClientErrorKind::BlobTooLarge(size) => ConcatBlobError::BlobTooLarge {
+                            blob_id: blob_id.to_string(),
+                            size: *size,
+                        },
+                        _ => ConcatBlobError::BlobReadFailed {
+                            blob_id: blob_id.to_string(),
+                            message: error.to_string(),
+                        },
+                    };
```

Comment:
Suggestion: Define a separate function for this or implement `From<ClientError> for ConcatBlobError` to avoid cluttering this function.

Reply:
Unfortunately From won't work because the errors can really benefit from the exterior context of the blob_id. I'll wrap this in a separate fn. Thanks.